### ISSUE TITLE
The reason for using rmsprop optimizer

### ISSUE THREAD

**Issue by Shaluols** [twni2016/Meta-SAC#1]
Hi,

Thanks for releasing the wonderful code. 
I wonder if there is a specific reason for using the RMsprop optimizer in the policy neural network (in the metasac.py file)? Because it doesn't work in my project if I use Adam as the policy optimizer. The log_aplha grad is always zero after the  `new_pg_loss.backward()` step.

**Comment by yufeiwang63** [comment]
Can you detail the changes you made? I think if the only change is to replace RMSprop with Adam at line 50 in sacmeta.py, log_alpha should still have non-zero grad after `new_pg_loss.backward()`, because when computing the `new_pg_loss`,  we are manually doing the RMSprop update for the policy network at line 136, and the benefit for doing so is that Pytorch can automatically compute the gradient of `new_pg_loss` w.r.t `log_alpha` for us. So simply changing the optimizer at line 50 should not affect this process.

We tried to use Adam for the policy update early (to do so, you should manually write a `simulate_adam` function that is similar to `simulate_rmsprop` at line 245 in metasac.py). The reason we disgard it is that, when using adam, the gradient of `new_pg_loss` w.r.t `log_alpha` often explodes to NaN. That's due to the different update rule Adam uses compared with RMSprop, it is less stable for this meta update process.

**Comment by Shaluols** [comment]
Hi, I didn't run your code but used almost the same architecture to learn the entropy coefficient. The changes I have made are the change of line 50 from RMSprop to Adam optimizer and replace the function `simulate_rmsprop` to `simulate_adam`. Here is my `simulate_adam` function:
```
def simulate_adam_step(self, optim):
        new_params = []
        for group in optim.param_groups:
            for p in group['params']:
                if p.grad is None:
                    continue
                grad = p.grad.data # p.grad.data
                state = optim.state[p]  #
                # State initialization
                if len(state) == 0:
                    state['step'] = 0
                    state['exp_avg'] = torch.zeros_like(p.data)
                    state['exp_avg_sq'] = torch.zeros_like(p.data)
                exp_avg, exp_avg_sq = state['exp_avg'], state['exp_avg_sq']
                beta1, beta2 = group['betas']
                state['step'] += 1
                exp_avg.mul_(beta1).add_(1 - beta1, grad)
                exp_avg_sq = exp_avg_sq * beta2 + (1 - beta2) * torch.mul(grad, grad)
                denom = torch.sqrt(exp_avg_sq) + group['eps']
                bias_correction1 = 1 - beta1 ** state['step']
                bias_correction2 = 1 - beta2 ** state['step']
                step_size = group['lr'] * math.sqrt(bias_correction2) / bias_correction1
                new_param = p - step_size * exp_avg / denom
                new_params.append(new_param)
        return new_params
```
What I found is that the grad of log_alpha is always zero after the `new_pg_loss.backward()` step, no matter I use the  `self.policy_optim.step()` or `simulate_adam_step` to update the policy parameters. However, when I change Adam to RMSprop optimizer and use the `simulate_rmsprop` function to update policy, it works well. 

So, I guess maybe you also encounter this problem and you change to RMSprop optimizer because of that.

**Comment by twni2016** [comment]
We didn't try Adam for policy optimizer, since we have some previous experience on self-written RMSProp for meta-gradient purpose. In theory Adam should work as it is like RMSProp with momentum. Please check the correctness of your Adam code, e.g. try without momentum.
